### PR TITLE
WT-8053 Update compact architecture guide page

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

**Comment by AlexeyAnisimovAU** [comment]
Do you think it makes sense to add `arch-compact` link to `arch-block.dox` (https://source.wiredtiger.com/develop/arch-block.html#Compaction)?

**Comment by AlexeyAnisimovAU** [comment]
`s_all` crashes with `/src/docs/arch-index.dox:60: error: unable to resolve reference to 'arch-compact' for \ref command (warning treated as error, aborting now)`.  Maybe the new page has to be added to `docs_data.py`?

File: src/docs/arch-compact.dox [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+that can be moved.
+
+Diving deep into the compaction process, the btree layer walks the internal pages for a given table to
+identify leaf pages that can be rewritten. For each leaf page referenced by the internal page, we use
+a block manager API to assess whether it will be beneficial to rewrite the leaf page. If the block
+manager recognizes that rewriting the page will be beneficial, it rewrites the disk
+block to a new disk location. The new disk address is placed into the parent internal page and
+the page is marked dirty. In this way the leaf page is rewritten without being brought into the
+cache. If the leaf page is already in cache, the leaf page is simply marked dirty and the subsequent
+checkpoint or eviction rewrites the page.
```
Comment (AlexeyAnisimovAU) [inline src/docs/arch-compact.dox:37]: We need to mention that compaction moves internal pages too. And the root page will be rewritten by the checkpoint. You can find comments about that in `__compact_walk_internal`.

File: src/docs/arch-compact.dox [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+referencing the new block used for the page. Once the checkpoint referencing the old block is deleted,
+the old disk block becomes available for reuse. The aim of compaction is to copy over the data from
+the blocks at the end of file to the blocks in the start of the file that are available for reuse.
+
+The compaction process starts with a system-wide checkpoint. There are potentially many dirty blocks
+in the cache, and the intention here is to write them out and then discard previous checkpoints so
+that there are as many blocks as possible on the file's "available for reuse" list when the compaction
+starts.  The compaction logic then starts examining the files that need to be compacted. Compacting
+the object is done 10% at a time, that is, compaction tries to move blocks from the last 10% of the
+file into the beginning of the file (the 10% is hard coded in the block manager).  The reason for
```
Comment (AlexeyAnisimovAU) [inline src/docs/arch-compact.dox:23]: Do you think the following information would be useful here?
`We then compact the high-level object, ignoring files smaller than 1MB. We compute the available bytes in the initial 80% and 90% of the file. If we can’t recover at least 1MB, the file is skipped. If at least 20% of the file is available in the first 80% of the file, we'll try compaction on the last 20% of the file. Else if at least 10% of the total file is available in the first 90% of the file, we'll try compaction on the last 10% of the file. Else the file is skipped.`

File: src/docs/arch-compact.dox [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+writes the checkpoint itself at the end of the file, which would prevent any file truncation.  When
+the metadata is updated for the second checkpoint, the blocks freed by compaction become available for
+the third checkpoint, so the third checkpoint's blocks are written towards the beginning of the file,
+and then the file can be truncated.
+
+Compaction is a time consuming operation and therefore logs have been added in compaction logic to
+track current progress. Some of the verbose logs for compaction can be activated by setting
+\c "verbose=[compact,compact_progress]" configuration option for \c wiredtiger_open() method. Apart from
+verbose logs, WT also records statistics about compact progress that can be used to debug issues
+with compaction process.
```
Comment (AlexeyAnisimovAU) [inline src/docs/arch-compact.dox:55]: Maybe we should mention `__VERBOSE_OUTPUT_PERCENTILE` definition that provides even more detailed verbose output?

**Review by AlexeyAnisimovAU — COMMENTED** [review]
Good job! It's coming together very nicely. I have a few questions/suggestions.

File: src/docs/arch-compact.dox (hunk 2 above)
Comment (raviprakashgiri29) [reply src/docs/arch-compact.dox:23]: Good point. Can we use the flowchart I created for the knowledge-sharing session?
Link to the presentation https://docs.google.com/presentation/d/17_vNBtBie3fm4kYRRS6CbBeMRRcTD5a7jSeQ34FDDw8/edit#slide=id.gf085a5dda9_0_18

Comment (hbokhari) [reply src/docs/arch-compact.dox:23]: I will go for the text option as making the flow chart might be a more complex task as we need to draw diagrams suing UML drawer.

File: src/docs/arch-compact.dox (hunk 3 above)
Comment (hbokhari) [reply src/docs/arch-compact.dox:55]: I would say `__VERBOSE_OUTPUT_PERCENTILE` is something that end user can not set through an API and therefore I would avoid mentioning this here. This doc is also intended for readers that do not deal with WT codebase.

File: src/docs/arch-compact.dox (hunk 2 above)
Comment (hbokhari) [reply src/docs/arch-compact.dox:23]: Added more details in 3d77c6f.

Comment (hbokhari) [reply src/docs/arch-compact.dox:29]: Added more details in 3d77c6f.

File: src/docs/arch-compact.dox (hunk 1 above)
Comment (hbokhari) [reply src/docs/arch-compact.dox]: I discuss this more in ce28d39.

**Comment by hbokhari** [comment]
@AlexeyAnisimovAU, I have added a reference to block manager page for the new compaction page.

**Review by raviprakashgiri29 — APPROVED** [review]
LGTM

File: src/docs/arch-compact.dox [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+compaction process involves rewriting disk blocks from the end of the on-disk file to the start of
+the file, hence giving an opportunity to reclaim the space from the end of the file through truncation
+and effectively reducing the file system reported file size. Note that the compaction is a
+"best-effort" process and does not guarantee a reduction in file size.
+
+Before understanding the compaction process, we need to understand how disk blocks are referenced by the
+checkpoints. Each checkpoint references a certain set of disk blocks for a table. When a dirty
+btree page is reconciled, a new disk block is assigned for the page and a new checkpoint starts
+referencing the new block used for the page. Once the checkpoint referencing the old block is deleted,
+the old disk block becomes available for reuse. Compaction process copies over the data from
```
Comment (AlexeyAnisimovAU) [inline src/docs/arch-compact.dox:15]: Do you think it would be more clear if we say here `Compaction process **moves** the data`?

**Review by AlexeyAnisimovAU — APPROVED** [review]
I like how you structured the information, well done! LGTM!

**Pull request merged**